### ISSUE TITLE
ISC affects post’s editing state without user input

### ISSUE THREAD

**Issue by Galanos** [image-source-control/image-source-control#172]
This one’s a bit strange and hard to both pinpoint and explain so I’ll add videos and a staging site with demo posts.

I had noticed for a while that WordPress had started to count a post as edited although I hadn’t done any modifications to it because it warned me to stay on the page even though I had just entered edit mode and tried to leave it again right away.

It turns out that the state of ISC source fields of images in the post has an effect on the post’s editing state.

This bug appears both in live 2.6.0 as well as in the current master branch.

@cnhk Again, @webzunft asked me to create this issue and assign it to you.

How to reproduce
---

There are two scenarios that I have explained in detail in the content of two posts and in two videos. @cnhk I’ll send you the link to the staging site with those posts and to my demonstrational videos in Slack.

Expected behavior
---

Depends on the nature of the observed behavior. Ideally, it would stop.

Priority
---

If there are no underlying complications yet unnoticed, this bug is just an annoyance and has low priority. But I think it should be checked out.

Additional context
---

I noticed and analyzed the bug while working on [issue 171](https://github.com/image-source-control/image-source-control/issues/171). @cnhk I don’t think there’s a connection, but please draw your own conclusions.

Tests
---

Again, the cause and the possible necessity of the observed behavior are too unclear to propose any tests.

**Comment by cnhk** [comment]
This is because how the ISC fields we add in core blocks work.

Like when using the same image multiple image block, when you edit the caption of one block, the post content is actually changed. The problem with ISC block is that, a block normally just loads with default values. But for ISC we need to get the image ID first, get the post meta for that image, then set the ISC field values for that image by calling `props.setAttributes` which for WordPress translates as "the post has been updated". It's valid for the public version as well as the one using REST API to load the values.

The values of a block's attributes is normally stored in the post content itself but we can't do that, the ISC values for the same image might be de-synced across multiple pages, or even on the same page like your demo. We can resynchronize the values on page load but it will result in calling the `props.setAttributes` function again.

I will try to see if there is a way to tell WordPress that we're not updating the post content when calling the `setAttributes` function.

**Comment by Galanos** [comment]
All right, I think I understand most of it. It’s annoying but nothing to worry about. I guess it’s up to @webzunft to decide whether we get into the trouble trying to fix this. For now, please don’t put too much time into your research until we’ve heard back from him:
> I will try to see if there is a way to tell WordPress that we're not updating the post content when calling the `setAttributes` function.

**Comment by Galanos** [comment]
I’ve returned from a trip and I’ve talked to @webzunft today. We agreed that this is too annoying to confront users with it. So I’d kindly ask you, @cnhk, to look into fixing this. Please get back to us if this takes you more than 2 to 3 hours before proceeding.
As described, this is already occurring in the current release, so it’s rather urgent. Thank you!

**Comment by cnhk** [comment]
@Galanos My first idea (tweaking `setAttributes`) looks like a dead end. This function is designed form the start to modify the block's properties.

So instead of using the components provided by WordPress, we have to create and use our own and place them where the current controls are. These components don't touch the block (image|cover|Media-Text) attributes but just load and save data from the image  post meta.

We need to create 3 components: one for text inputs (with all the labels, description etc), one select input, and a checkbox. But that will probably take more than 3 hours. 4 or may be 5 not sure yet.

**Comment by webzunft** [comment]
I am in a few developer channels where I could also ask for best practices, advice, or maybe plugins that do this and from which we can copy. @cnhk does that make sense? If so, could you write me 1-2 sentences that I can just post there?

**Comment by cnhk** [comment]
@webzunft Something like:

- What is the best way for adding inputs to core blocks that only read and write post meta without creating/modifying any block attribute?

**Comment by cnhk** [comment]
It seems that updating a Gutenberg block field (update the displayed value) without using `setAttributes` won't work unless using a hack (which I didn't found/searched yet).

If we're going to fix it, I would suggest a totally different approach.

- Adding a single block field (a WPElement) that provide us the image ID
- Handling the display/update by ourselves with just JS (no react/WPElement/Gutenberg). So no matter what happens on the ISC fields, it has no effect on the post state.

Waiting for feedback from @webzunft in cases someone has a solution for this.

**Comment by webzunft** [comment]
> What is the best way for adding inputs to core blocks that only read and write post meta without creating/modifying any block attribute?

@cnhk This is what I got from one developer. Does it help?

> Make sure the meta value is registered with show_in_rest (or it'll just disappear into a black hole), and then inject it in the sidebar I'd say.
> You can see how the [Block Visibility](https://wordpress.org/plugins/block-visibility/) plugin does the injection, it is one of the smoothest approaches to adding very custom data to existing blocks that I have worked with so far

**Comment by cnhk** [comment]
The JS files of [Block Visibility](https://wordpress.org/plugins/block-visibility/) are all minified/mangled (uglifyJs) but I grasped more or less the idea. This is roughly the same thing we do with our custom REST API calls, but using the WordPress [Core Data methods](https://developer.wordpress.org/block-editor/reference-guides/data/data-core/).

It's about how to fetch and save post meta from within a block's code. But there is the same problem. The display isn't updated without calling `setAttributes`. The fields of **Block Visibility** are all attached to a block attribute (when you touch one of them, the post is considered to be edited). It's just that there are stuffs that they save in post meta.

For example, if you have an image source "xy", you select that field and type "z". At this point we can update the internal Gutenberg's core data, even update the post meta, but the text field won't display "xyz" unless `setAttributes` is called if we're using a `wp.components.TextControl` (the WP's text control).

**Comment by webzunft** [comment]
@cnhk thanks for looking into it!
Do I understand correctly, that you are about to solve the issue that the saving-popup appears all the time, and instead it only shows when the ISC fields are in fact changed? That is ok. It seems to correspond with what WordPress core has in mind.
Do we then even need the REST API call that stores changed values in our fields or can that be removed in favor of storing the data when the post is saved?

**Comment by cnhk** [comment]
> solve the issue that the saving-popup

Yes.

> Do we then even need the REST API call that stores changed values in our fields or can that be removed in favor of storing the data when the post is saved?

- The REST API call that load the values can be replaced by the one provided by the core mechanism (that will actually avoid the save pop-up)
- Right now, we are using a standard AJAX call `admin-ajax.php` to save the changes, not the REST API. There is also a "save meta" feature in the core function, But I didn't checked yet if it sends it via a standalone REST API call or only when saving the entire post (logically should be the later but I'm going to see what happens anyway)

**Comment by cnhk** [comment]
Waiting for [this PR](https://github.com/image-source-control/image-source-control/pull/181) to be merged. Don't want to resolve other conflicts later.
